server/plpgsql/statements.go

```diff
@@ -0,0 +1,206 @@
+
+// AppendOperations implements the interface Statement.
+func (stmt Block) AppendOperations(ops *[]InterpreterOperation, stack *InterpreterStack) {
+	stack.PushScope()
+	*ops = append(*ops, InterpreterOperation{
+		OpCode: OpCode_ScopeBegin,
+	})
+	for _, variable := range stmt.Variable {
+		if !variable.IsParameter {
+			*ops = append(*ops, InterpreterOperation{
+				OpCode:      OpCode_Declare,
+				PrimaryData: variable.Type,
+				Target:      variable.Name,
+			})
+		}
+		stack.NewVariableWithValue(variable.Name, nil, nil)
+	}
+	for _, innerStmt := range stmt.Body {
+		innerStmt.AppendOperations(ops, stack)
+	}
+	*ops = append(*ops, InterpreterOperation{
+		OpCode: OpCode_ScopeEnd,
+	})
+	stack.PopScope()
+}
+
+// Goto jumps to the counter at the given offset.
+type Goto struct {
+	Offset int32
+}
+
+var _ Statement = Goto{}
+
+// OperationSize implements the interface Statement.
+func (Goto) OperationSize() int32 {
+	return 1
+}
+
+// AppendOperations implements the interface Statement.
+func (stmt Goto) AppendOperations(ops *[]InterpreterOperation, stack *InterpreterStack) {
+	*ops = append(*ops, InterpreterOperation{
+		OpCode: OpCode_Goto,
+		Index:  len(*ops) + int(stmt.Offset),
+	})
+}
+
+// If represents an IF condition, alongside its Goto offset if the condition is true.
+type If struct {
+	Condition  string
+	GotoOffset int32
+}
+
+var _ Statement = If{}
+
+// OperationSize implements the interface Statement.
+func (If) OperationSize() int32 {
+	return 1
+}
+
+// AppendOperations implements the interface Statement.
+func (stmt If) AppendOperations(ops *[]InterpreterOperation, stack *InterpreterStack) {
+	condition := stmt.Condition
+	var referencedVariables []string
+	for varName := range stack.ListVariables() {
+		if strings.Contains(condition, varName) {
+			referencedVariables = append(referencedVariables, varName)
+		}
+		condition = strings.Replace(condition, varName, fmt.Sprintf("$%d", len(referencedVariables)), 1)
+	}
+	*ops = append(*ops, InterpreterOperation{
+		OpCode:        OpCode_If,
+		PrimaryData:   "SELECT " + condition + ";",
+		SecondaryData: referencedVariables,
+		Index:         len(*ops) + int(stmt.GotoOffset),
+	})
+}
+
+// Return represents a RETURN statement.
+type Return struct {
+	Expression string
+}
+
+var _ Statement = Return{}
+
+// OperationSize implements the interface Statement.
+func (Return) OperationSize() int32 {
+	return 1
+}
+
+// AppendOperations implements the interface Statement.
+func (stmt Return) AppendOperations(ops *[]InterpreterOperation, stack *InterpreterStack) {
+	// TODO: figure out how I'm supposed to actually do this, rather than a search and replace
+	expression := stmt.Expression
+	var referencedVariables []string
+	for varName := range stack.ListVariables() {
+		if strings.Contains(expression, varName) {
+			referencedVariables = append(referencedVariables, varName)
+		}
+		expression = strings.Replace(expression, varName, fmt.Sprintf("$%d", len(referencedVariables)), 1)
+	}
```

Comment:
Looks like we've got this code block duplicated three times already and seems like it's going to be useful for other statements, so seems worth pulling out into a shared function.

Reply:
I didn't bother doing so since it's known incorrect. As we discussed over DMs though, it looks like we may have another way to handle this, so we'll be replacing all of this anyway.